Feature/jetpack connection flow

planarvoid opened the pull request:
Open Jetpack web connection flow when users tries to open Stats and Jetpack is not connected. Replace standard WordPress.com login with the in-app version. Redirect the result back to the stats using a deeplink. 

To test:
## Selfhosted site without Jetpack
* Login with self-hosted site without Jetpack and WP.com 
* Click on Stats
* Jetpack connection flow opens in Web view
* Follow the steps
* If user is not connected to WP.com
  * Flow prompts user to login to WP.com account 
  * Opens new Login process in the app
  * Redirects user back to the flow
* User approves the Jetpack connection
* User is redirected to the Stats

nbradbury commented:
I don't think we want any of the icons (including the options menu) here.

![screenshot_1517339899](https://user-images.githubusercontent.com/3903757/35586568-2c5067c2-05c9-11e8-8875-187aa57d4734.png)

nbradbury commented:
When I get to the step where I login to wp.com, it says "Log in to WordPress.com to access the **post** "

![screenshot_1517341892](https://user-images.githubusercontent.com/3903757/35588081-96838b70-05cd-11e8-8b11-4fa8053dc39f.png)

nbradbury commented:
I seem to have gotten the app in a weird state. I click on stats, click YES on the "Jetpack plugin not found" dialog, then after the webview loads I get returned to the main screen which shows a toast stating "already-connected" (when in fact Jetpack is not connected).

Clearing storage and logging in again fixed the problem.

![screenshot_1517343086](https://user-images.githubusercontent.com/3903757/35591804-2dc022b8-05d9-11e8-937b-7ba0220cea35.png)

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/JetpackConnectionWebViewClient.java:
```diff
@@ -0,0 +1,137 @@
+    public boolean shouldOverrideUrlLoading(WebView view, String stringUrl) {
+        try {
+            final Uri url = Uri.parse(stringUrl);
+            final SiteModel site = mSiteStore.getSiteByLocalId(AppPrefs.getSelectedSite());
```
Elsewhere in the app we pass the entire site model using `WordPress.SITE` (see [ActivityLauncher](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/ActivityLauncher.java#L101) for examples).

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/JetpackConnectionWebViewClient.java:
```diff
@@ -0,0 +1,137 @@
+            }
+        } catch (UnsupportedEncodingException e) {
+            AppLog.e(AppLog.T.API, "Unexpected URL encoding in Jetpack connection flow.", e);
+            e.printStackTrace();
```
`AppLog.e` already logs the stack trace, so `e.printStackTrace()` isn't necessary here.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/JetpackConnectionWebViewClient.java:
```diff
@@ -0,0 +1,137 @@
+    private static final String LOGIN_PATH = "/wp-login.php";
+    private static final String ADMIN_PATH = "/wp-admin/admin.php";
+    private static final String REDIRECT_PARAMETER = "redirect_to=";
+    private static final int REQUEST_CODE = 1;
```
Although it's not necessary to change this, I wanted to point out [RequestCodes.java](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/RequestCodes.java) which contains most of the app's request codes.

nbradbury commented on libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTracker.java:
```diff
@@ -186,6 +186,8 @@
         PERFORMED_JETPACK_SIGN_IN_FROM_STATS_SCREEN,
         STATS_SELECTED_INSTALL_JETPACK,
+        STATS_CANCELED_INSTALL_JETPACK,
+        STATS_COMPLETED_INSTALL_JETPACK,
```
It doesn't look like `STATS_COMPLETED_INSTALL_JETPACK` is ever used.

planarvoid replied:
Oh I used a wrong value for completed event, fixed now 👍

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/JetpackConnectionWebViewClient.java:
```diff
@@ -0,0 +1,137 @@
+    private boolean flowFinished = false;
+
+    JetpackConnectionWebViewClient(Activity activity, AccountStore accountStore, SiteStore mSiteStore) {
+        this.activity = activity;
```
Rather than pass the stores as parameters, we can just use:

```
@Inject AccountStore mAccountStore;
@Inject SiteStore mSiteStore;
```

planarvoid replied:
The JetpackConnectionWebViewClient is not an injected field i the activity. I can create an injected factory in order to inject the params. Is that a pattern we use?

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/JetpackConnectionWebViewClient.java:
```diff
@@ -0,0 +1,137 @@
+        try {
+            final Uri url = Uri.parse(stringUrl);
+            final SiteModel site = mSiteStore.getSiteByLocalId(AppPrefs.getSelectedSite());
+            final String loadedHost = url.getHost();
```
I think we need to add some null-checking on `loadedHost` and `loadedPath` because there are cases where `getPath()` and `getHost()`  can return null. It's a rare possibility here, but since the user can click on numerous external links during this flow it's worth checking for.

nbradbury commented:
@planarvoid I'm done with my first round of reviews. Really nice work on this one - certainly not an easy task to tackle right after joining us!

planarvoid commented:
@nbradbury I moved the Jetpack logic to a new child activity (to make it more readable) and fixed most of the comments. 
2 questions: 
* How did you get the weird state? It's a bit hard to get all the usecases when I always have to recreate the site 😢 
* How did you get to the login flow with the wrong text? I'm seeing this: 
![screenshot_1517390916](https://user-images.githubusercontent.com/1079756/35619138-508b6c04-067d-11e8-8427-19eee028fd87.png)

nbradbury commented:
> How did you get the weird state? It's a bit hard to get all the usecases when I always have to recreate the site

I was able to repro this but not consistently. Start the app with a site that has Jetpack activated, then go to the web app and de-activate Jetpack. The next time you tap Stats in the app, it *should* correctly tell you that Jetpack needs to be enabled, but then the web view exits with that toast.

> How did you get to the login flow with the wrong text? I

I can repro that consistently by logging out of wp.com in the app, leaving only the self-hosted site available. I then enter the Jetpack flow, and when I'm asked to approve the connection to wp.com it takes me to that screen.

planarvoid commented:
@nbradbury I fixed the second issue (it was happening when the user was logged in to the wpcom in the web view but not in the app. 
The first issue was discussed off GH.

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/stats/StatsDeeplinkActivity.java:
```diff
@@ -100,19 +104,17 @@ protected void onStart() {
         super.onActivityResult(requestCode, resultCode, data);
         // show the post if user is returning from successful login
-        if (requestCode == RequestCodes.DO_LOGIN && resultCode == RESULT_OK) {
+        if (requestCode == RequestCodes.JETPACK_LOGIN) {
```
Do we know for certain that the login was successful here? Also, I think the above comment is a copy & paste mistake - "show the post" doesn't apply to this situation.

planarvoid replied:
We probably don't, I'll add some "if not ok, finish"

nbradbury commented:
Looks good, great work! :shipit:
